### ISSUE TITLE
Imprint quality

### ISSUE THREAD

**Issue by axi92** [cadon/ARKStatsExtractor#59]
I got a bird with this stats now?
And the tool is confused obviously xD
![mutant](https://cloud.githubusercontent.com/assets/1394837/15940220/4fd8c4a4-2e7b-11e6-995f-ef131fadfc7a.jpg)

**Comment by Capps0611** [comment]
Hey so, We had a Ptera that we bred, she hatched at lvl 220 with 14% imprinting. I only just now started using your tool, The one stat I did some math on was her carry wait. (due to no domesticated level ups being assigned to it)

Pre-imprinting carry weight: 270.0
Post-Imprinting carry weight: 277.7
Rough Percentage Increase: 2.85%

I also ran the numbers with the health (also no domesticated level ups applied)  after I found the 2.85% increase with the carry weight and found the same 2.85% increase. Unfortunately I don't have any other numbers to compare to because after many level ups I don't remember what her other base stats were, however I hope this info can help you!

**Comment by cadon** [comment]
Thanks for the data. Currently we think the imprinting-bonus is 20% of the stat when hatched (multiplied by the imprinting-bonus-percentage). The issue still is, that some creatures seem to gain a bonus in all stats, other only in a few, we don't know yet on what that depends.

**Comment by VolatilePulse** [comment]
I figured I would add some new information to this request since it is already opened. I was doing some research today about imprinting and the stats associated with it and stumbled across this link https://survivetheark.com/index.php?/forums/topic/65370-baby-dino-interactions/&page=3

From what I was reading, I haven't been able to determine 100% yet, it looks like the stat boost from imprinting is directly linked to the final imprinting percentage as well. I did read in a few places that the stat boost is unaffected if the percentage drops though. For example, if I get a 50% imprint half way through, will the stats be persistent through the breeding or will it drop in correlation to the percentage. If hp is increased by 10% will it go back to 0% increase if imprint drops to 0% as well or will the stats remain unchanged? That statistic alone will be the determining factor on if I interact with all of my babies or only my finaly premium bird. I also could charge extra money since that extra boost in stats in time consuming and results in a better breed.

I hope that thread can help answer a lot of your questions for adding this feature into your extractor and I wish you luck. I have a feeling it's going to get a bit hairy with the new calculations.

**Comment by cadon** [comment]
Hi, thanks for the link, that confirmed my findings so far. It's still strange that some cuddles not give a bonus to all the usual stats. Maybe that's a bug.

- AFAIK the imprinting-bonus is not inherited, so you don't have to cuddle the non-final creatures.
- I've not yet tested if the stat-values decrease again, if the bonus resets, but I guess they will. Maybe only after a server-restart (only guessing here).

**Comment by NakramR** [comment]
If your dino has 50% imprinting bonus (and therefore a 10% increase in HP)
and you stop caring for it, the HP bonus decreases back to 0%

**Comment by VolatilePulse** [comment]
What I have read so far is that the imprinting bonus is NOT carried on to next generation dinos so it is inefficient for breeding a perfect breed. I also noticed that it was mentioned that the stat increases happen during every imprint increase, it just doesn't reflect it immediately. One poster mentioned that logging out and back in tended to remedy the stats not adjusting. I haven't started with imprinting yet since it's already rough raising 5+ argies as a solo tribe on an official server, but I plan on doing so next week.

It was also mentioned that all stats, including speed and torpor, are increased with the exception of stamina and oxygen. @NakramR That makes it worthless for me to attempt wasting kibble for sell imprinted dinos then, and makes this that much harder. Basically, is it more rewarding to just interact with them for the last day as opposed to trying interact whenever you can (I like my sleep and my job so I'm not willing to give either up for little to no gain, in the event it reaches 0% again)? I assumed the imprinting percentage was only for the rider bonus, but the direct correlation to the stat increase is a bummer.

**Comment by loginator** [comment]
I tested imprinting with the new update (Rex, Carno, Ptera, Argi) and it just increased HP of the dinos.
Can you guys confirm that?
Or can you give me an example where this is not the case?

**Comment by cadon** [comment]
You mean with the new game-update, and the increase of only the HP was ingame, too?

The game does not increase all stats everytime you do an imprinting-action, sometimes it apparently "forgets" to increase some stats (keep in mind that Stamina and Oxygen are never affected by imprinting). But after another imprinting-action it updates the forgotten stats as well, so in the end it doesn't seem to matter. Maybe the forgotten stats are updated after a server-restart as well.

If you mean the tool has an error, describe the issue more precisely. In general if you find a new bug, open a separate issue-thread for that. For discussing the game, better use the forums or reddit, as there are more people who possibly have answers.

**Comment by loginator** [comment]
No not an error I just wanted to share my tests :)
I tested this on a private session and used the SetImprintQuality command.
After setting the quality to 100% only health increased by 20% in all cases.

**Comment by jimethn** [comment]
So the way imprinting affects the stats is listed in the wiki here: http://ark.gamepedia.com/Creature_Stats_Calculation

Are you saying this math is wrong, or doesn't apply to every stat?

**Comment by cadon** [comment]
The math is correct and not very hard. The problem is, the game doesn't upgrade all stats each time a cuddle (or other imprinting-action) is given. I guess that's a bug, as after some time or a server-restart it seems to be correct.

That alone makes it already hard to get the correct numbers, but the fact that the game uses a precision of two decimals internally and only shows rounded integers for the current imprinting-process, multiplies the possible level-combinations by 100. It would make an extraction impractical, as the user had to choose the correct combination from a list of some hundred possibilities, without knowing which one is the correct one.

Maybe there's a way to simplify that, and if I can think of that, I'll implement it.

**Comment by Remoraz** [comment]
As far as the stats updating - seems to be a client side caching issue of some sort. A relog will fix it 100% of the time, as will upping that stat, and sometimes upping any stat will fix them all - though, I know we can't code to "sometimes".

Had a thought on deciphering the stats. Since we're primarily concerned with the base stats for breeding, would it be impractical to require the parents of an imprinted extraction already be known? If that were the case, the number of guess would be Mother/Father +(0-2) for mutation. Seems like some logic for stat matching already exists in the drop downs. (If you did go this route, you could also determine any mutations based on a known mother/father stat set and having a stat that's correlated to neither.)

**Comment by Zhinjio** [comment]
Could there be a separate extract for the ancestor's screen that would figure out mutations?

**Comment by cadon** [comment]
@Remoraz choosing the parents first is a good idea, at least for the cases the parents are known.
@Zhinjio do you mean a tab similar to the "Pedigree"-tab that displays which stats are mutated? Sounds good as well.
If you have more ideas or want to clarify them, better create a new issue, so things don't get lost in this imprinting-thread.

**Comment by cassfalg** [comment]
I had an idea to reduce the number of possible combinations due to the rounding issues with the UI, or rather: the actual combinations are not as huge as it seems. As far as I know, the boost from each imprint is fixed for each creature, something like total maturation time divided by number of possible imprints times 100. There is only a finite amount of exact values that the imprint quality can have.

This likely gets more complicated on unofficial servers, as it depends on the according settings, but it should still be possible to derive the exact imprint values from the displayed rounded ones. Thus, if someone enters 74% imprint on a Pteranodon there is likely only one single exact value within rounding distance, and the correct imprint with decimals can be found, reducing the possible combinations.

Or, in other words: Depending on the Breeding related settings, the possible exact values for the imprint are limited and known. Instead of relying on the inexact input, you could chose the exact value from the possible combinations, or simply make a dropdown list for entering imprint quality.

Shouldn't that reduce the possible combinations enough to make extraction possible?

**Comment by cadon** [comment]
If that info is correct it would be great and definitively help! Do you have any source with tables of these numbers or how they are calculated?

**Comment by cassfalg** [comment]
No, I'm sorry. I just remembered reading about this a couple months back...

The best I found is from the Wiki (http://ark.gamepedia.com/Imprinting):
"Each care session will increase Imprinting affinity by a set amount, based on total maturation time. It is approximately: Imprinting Affinity Gain = 4 Hours / Total Maturation Time of Species. The displayed value is rounded from the actual value, thus, for a Pteranodon, which is ~10.8%, it displays as 11% per care, but after 9 care intervals you will have 97% (actually 97.2%)."

Sounds about right, though I did not run any numbers myself.

**Comment by cadon** [comment]
@cassfalg wow, that actually works! That was a great idea and pretty simple to implement, thanks!

The game sometimes has hickups, so that it "forgets" to increase the values after a caring-action, after a game-restart or server-restart that should be resolved, though.

I'll upload a new release soon.
